### ISSUE TITLE
The serialization cliff

### ISSUE THREAD

**Issue by mahmoud** [ericvsmith/dataclasses#20]
Seeing as the scope of dataclasses still seems fairly pliable, I'll share an answer to #19 ("why not attrs?") which could greatly increase its utility: The serialization cliff.

My teammates and I have been using attrs almost as long as it's been around, and namedtuples for much longer. Great and fine solutions, if your data doesn't have to leave the process space.

The boilerplate we're trying to avoid comes back in a severe way as soon as databases, SOA, or files get involved. Let alone anything with complicated validation rules.

I think to be true to the name, dataclasses need to account for the pervasiveness of data ingestion and emission. As it stands, attrs has `asdict()` and not much else. Nothing to help recursively (re)construct instances from data.

There is another, somewhat more popular, approach in [marshmallow](https://github.com/marshmallow-code/marshmallow), which also aims to make data-centric, serialization-agnostic container types. It's not perfect, but it's a starting point. A balance of attrs and marshmallow features may yield a powerful and sufficiently-differentiated featureset that I think a lot of developers are missing in Python. The number of entries in this space seems to agree.

I'm happy to see more discussion going into these fundamentals, and I'm hopeful that more and more they take the whole data workflow into mind. :)

**Comment by ericvsmith** [comment]
@mahmoud: Can you summarize marshmallow's approach?

**Comment by mahmoud** [comment]
I may not be the best person to try, but here are the parts that jump out to me:

1. Explicit nesting of container types, through [`Nested` fields](http://marshmallow.readthedocs.io/en/latest/nesting.html). (This contrasts with attrs, whose `.asdict` has a recursive option that relies on value introspection)
2. An adapter [ecosystem](https://github.com/marshmallow-code/marshmallow/wiki/Ecosystem) (and an API that enables it), so that translation/serialization approaches can be externalized and shared.

Other notable aspects I'm not as eager to preserve: 

1. The inheritance-driven approach (inheriting from `Schema`).
2. The separation of schema and container. Marshmallow works by serializing vanilla dictionaries and other objects handed to the schema's `dump` and `load` methods.

The last point is what enables Marshmallow to interop with attrs and other similar libraries, but it also leads to a fair amount of definition duplication in my limited experience.

**Comment by mahmoud** [comment]
Also, I didn't intend to be too marshmallow-centric, another, maybe more-relevant, point of prior art would be [cattrs](https://pypi.python.org/pypi/cattrs). I believe it's Python 3-only and leverages new language features more.

**Comment by ilevkivskyi** [comment]
> Seeing as the scope of dataclasses still seems fairly pliable, I'll share an answer to #19 

I would put that differently. In my view, an important property that dataclasses should have is being easily extendable. They should provide a common way of adding functionalities that would otherwise be added via some boilerplate, for example, the serialization code. Importantly, such extensions should be easily _composable_ (i.e. should not interfere with each other). Since Python 3.6 we have several additional ways of customizing classes (such as ``__init_subclass__``) that allow avoiding metaclass conflicts.

We don't need to include all functionality in stdlib (although some serialization support might be OK), but it should be possible to do something like this:
```python
from third_party import data_serializer

@data(serializer=data_serializer)
class C:
    x: List[int] = field(factory=list)
```
I am still not sure how exactly to organize such composable extensions, it looks we need to preserve the original namespace (that contains ``field(...)`` values) so that it is accessible by the extensions.

**Comment by ericvsmith** [comment]
I'm not sure why dataclasses needs to address serialization. It shouldn't add any features, compared to "regular" classes, that hinder any serialization you want to do.

Now, if you're saying that it could help serialization by making it easier to walk the hierarchy of nested dataclasses, then maybe we could do something there. But I don't understand the space well enough to have any suggestions.

**Comment by gvanrossum** [comment]
We should probably at least promise (and verify in tests) that pickling
will work without extra work, i.e. if all the field types are picklable,
the instances of the class should be picklable.

**Comment by ericvsmith** [comment]
I'll leave this issue open until those tests are added.

**Comment by ilevkivskyi** [comment]
@ericvsmith 
> I'm not sure why dataclasses needs to address serialization. It shouldn't add any features, compared to "regular" classes, that hinder any serialization you want to do.

I could imagine people might want some _basic_ support out of the box, but I think ``asdict()`` is already enough (especially if we add flags like ``nested`` and ``copy_values``).

**Comment by ericfrederich** [comment]
I agree the next logical step after dataclasses is serialization.
The next step after that is to throw in an ORM.

Eventually Django will be part of the stdlib.

**Comment by ericvsmith** [comment]
@ericfrederich : It's hard to gauge the tone of your message. I assume the last 2 lines are sarcastic. Does that therefore mean in the first line that you think dataclasses should not support serialization?

**Comment by mahmoud** [comment]
Haha, that certainly was my interpretation. Luckily for Mr. Frederich, that particular slope of Python's has not been slippery for quite a while.

But to go along with what Guido was saying (if all fields' values are picklable, the dataclass instance is picklable), why not make it such that if all fields' values are JSON-serializable, the dataclass instance is JSON serializable. 

That alone would be a massive gain on a long-unsolved problem in Python.

**Comment by ericfrederich** [comment]
@ericvsmith, @mahmoud

Sorry for confusion.  Only last sentence was sarcasm.  It has been my personal experience that each thing leads to the other.  I brought it up on this thread because if there is a *serialization* cliff to declaring attributes I likewise believe there to be a *orm* cliff to serialization.

It's a hard decision where to draw the line on all this *bolts-included* stuff

**Comment by ericvsmith** [comment]
I'm going to close this. I think that per-field metadata (see #69) addresses this issue. I can't think of a reason you'd need to have per-class metadata to address serialization, but maybe you would. I think https://github.com/ericvsmith/dataclasses/issues/20#issuecomment-311901690, if needed, would be addressed by per-class metadata instead of a `serializer` on the dataclass.

I'm not completely opposed to adding per-class metadata, if needed.
